File: money-java-servlet/src/main/scala/com/comcast/money/java/servlet/TraceFilter.scala
/*
 * Copyright 2012 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.comcast.money.java.servlet

import com.comcast.money.core.Money
import io.opentelemetry.context.{ Context, Scope }
import javax.servlet._
import javax.servlet.http.{ HttpServletRequest, HttpServletRequestWrapper, HttpServletResponse }
import org.slf4j.LoggerFactory

/**
 * A Java Servlet 2.5 Filter.  Examines the inbound http request, and will set the
 * trace context for the request if the money trace header or X-B3 style headers are found
 */
class TraceFilter extends Filter {

  private val logger = LoggerFactory.getLogger(classOf[TraceFilter])
  private val tracer = Money.Environment.tracer
  private val formatter = Money.Environment.formatter

  override def init(filterConfig: FilterConfig): Unit = {}

  override def destroy(): Unit = {}

  private val spanName = "servlet"

  override def doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain): Unit = {

    val httpRequest = new HttpServletRequestWrapper(request.asInstanceOf[HttpServletRequest])

    val scope: Scope = formatter.fromHttpHeaders(httpRequest.getHeader, logger.warn) match {
      case Some(spanId) =>
        val span = tracer.spanFactory.newSpan(spanId, spanName)
        Context.root()
          .`with`(span)
          .makeCurrent()
      case None => () => ()
    }

    try {
      val httpResponse = response.asInstanceOf[HttpServletResponse]
      formatter.setResponseHeaders(httpRequest.getHeader, httpResponse.addHeader)

      chain.doFilter(request, response)
    } finally {
      scope.close()
    }
  }

}

File: money-java-servlet/src/main/java/com/comcast/money/java/servlet/TraceFilter.java
/*
 * Copyright 2012-2015 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.comcast.money.java.servlet;

import java.io.IOException;
import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletRequestWrapper;
import javax.servlet.http.HttpServletResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.comcast.money.core.Money;
import com.comcast.money.core.SpanId;
import com.comcast.money.core.Tracer;

public class TraceFilter implements Filter {

    private static final String MONEY_TRACE_HEADER = "X-MoneyTrace";
    private static final Logger logger = LoggerFactory.getLogger(TraceFilter.class);

    protected Tracer tracer;

    @Override
    public void init(FilterConfig filterConfig) throws ServletException {

        tracer = Money.tracer;
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {

        HttpServletRequestWrapper httpRequest = new HttpServletRequestWrapper((HttpServletRequest)request);
        String moneyHttpHeader = httpRequest.getHeader(MONEY_TRACE_HEADER);

        if (moneyHttpHeader != null) {
            try {
                SpanId incomingSpanId = SpanId.fromHttpHeader(moneyHttpHeader);

                if (response instanceof HttpServletResponse) {
                    HttpServletResponse httpResponse = (HttpServletResponse)response;
                    httpResponse.addHeader(MONEY_TRACE_HEADER, moneyHttpHeader);
                }
                tracer.setTraceContext(incomingSpanId);
            } catch(Throwable ex) {
                logger.warn("Unable to parse money trace for request header '{}'", moneyHttpHeader);
            }
        }

        chain.doFilter(request, response);
    }

    @Override
    public void destroy() {

    }
}
